packages/salesforcedx-vscode-soql/src/client/codeCompletion.ts

```diff
@@ -0,0 +1,320 @@
+import { retrieveSObject, retrieveSObjects } from '../sfdx';
+
+import { Middleware } from 'vscode-languageclient';
+import { telemetryService } from '../telemetry';
+
+const EXPANDABLE_ITEM_PATTERN = /__([A-Z_]+)/;
+
+export const middleware: Middleware = {
+  // The SOQL LSP server may include special completion items as "placeholders" for
+  // the client to expand with information from the users' default Salesforce Org.
+  // We do that here as middleware, transforming the server response before passing
+  // it up to VSCode.
+
+  provideCompletionItem: async (document, position, context, token, next) => {
+    const items = (await next(
+      document,
+      position,
+      context,
+      token
+    )) as ProtocolCompletionItem[];
+
+    return expandPlaceholders(await filterByContext(items));
+  }
+};
+
+interface SoqlItemContext {
+  sobjectName: string;
+  fieldName?: string;
+  notNillable?: boolean;
+}
+
+async function filterByContext(
+  items: ProtocolCompletionItem[]
+): Promise<ProtocolCompletionItem[]> {
+  const filteredItems: ProtocolCompletionItem[] = [];
+
+  for (const item of items) {
+    if (
+      !EXPANDABLE_ITEM_PATTERN.test(item.label) &&
+      item?.data?.soqlContext?.sobjectName
+    ) {
+      const objMetadata = await retrieveSObject(
+        item.data.soqlContext.sobjectName
+      );
+      const fieldMeta = objMetadata.fields.find(
+        field => field.name === item.data.soqlContext.fieldName
+      );
+      if (fieldMeta && !operatorsByType[fieldMeta.type].includes(item.label)) {
+        continue;
+      }
+    }
+
+    filteredItems.push(item);
+  }
+
+  return filteredItems;
+}
+
+async function expandPlaceholders(
+  items: ProtocolCompletionItem[]
+): Promise<ProtocolCompletionItem[]> {
+  const expandedItems = [...items];
+
+  for (const [index, item] of items.entries()) {
+    const parsedCommand = item.label.match(EXPANDABLE_ITEM_PATTERN);
+    if (parsedCommand) {
+      const commandName = parsedCommand[1];
+
+      const handler = expandFunctions[commandName];
+      if (handler) {
+        expandedItems.splice(
+          index,
+          1,
+          ...(await handler(item?.data?.soqlContext))
+        );
+      } else {
+        telemetryService.sendException(
+          'SOQLLanguageServerException',
+          `Unknown SOQL LSP completion command ${commandName}!`
+        );
+      }
+    }
+  }
+
+  return expandedItems;
+}
+
+const expandFunctions: {
+  [key: string]: (
+    soqlContext?: SoqlItemContext
+  ) => Promise<ProtocolCompletionItem[]>;
+} = {
+  SOBJECTS_PLACEHOLDER: async (
+    soqlContext?: SoqlItemContext
+  ): Promise<ProtocolCompletionItem[]> => {
+    try {
+      const sobjectItems = (await retrieveSObjects()).map(objName => {
+        const item = new ProtocolCompletionItem(objName);
+        item.kind = CompletionItemKind.Class;
+        return item;
+      });
+
+      return sobjectItems;
+    } catch (metadataErrors) {
+      return [];
```

Comment:
If `retrieveSObjects()` throws an error, seems like something we might want to report

Reply:
Thanks. Done.